File: src/tqi/scoring/time_profile.py
"""Time-of-day TQI profile."""

import numpy as np

from tqi.config import DEPARTURE_TIMES, TSR_CAR, TSR_WALK


def compute_time_profile(
    per_slot_coverage: np.ndarray,
    per_slot_mean_tsr: np.ndarray,
) -> list[tuple[str, float]]:
    """Compute TQI for each time slot.

    Returns list of (time_label, tqi_score) tuples.
    """
    results = []
    for i, t_min in enumerate(DEPARTURE_TIMES):
        h, m = divmod(t_min, 60)
        label = f"{h:02d}:{m:02d}"

        cov_score = float(per_slot_coverage[i]) * 100
        tsr = float(per_slot_mean_tsr[i])
        spd_score = float(np.clip((tsr - TSR_WALK) / (TSR_CAR - TSR_WALK) * 100, 0, 100))
        slot_tqi = 0.5 * cov_score + 0.5 * spd_score

        results.append((label, slot_tqi))
    return results

File: go/internal/scoring/time_profile.go
package scoring

import (
	"fmt"
	"math"

	"github.com/jasondyck/chwk-tqi/internal/config"
)

// ComputeTimeProfile returns the TQI for each departure time slot.
func ComputeTimeProfile(perSlotCoverage, perSlotMeanTSR []float64) []TimeSlotScore {
	departures := config.DepartureTimes()
	n := len(perSlotCoverage)
	if n > len(departures) {
		n = len(departures)
	}

	profile := make([]TimeSlotScore, n)
	for i := 0; i < n; i++ {
		cov := perSlotCoverage[i] * 100.0

		speedNorm := (perSlotMeanTSR[i] - config.TSRWalk) / (config.TSRCar - config.TSRWalk) * 100.0
		speedNorm = math.Max(0, math.Min(100, speedNorm))

		tqi := 0.5*cov + 0.5*speedNorm

		h := departures[i] / 60
		m := departures[i] % 60
		label := fmt.Sprintf("%02d:%02d", h, m)

		profile[i] = TimeSlotScore{Label: label, Score: tqi}
	}
	return profile
}
